model_analyzer/monitor/nvml.py

```diff
@@ -0,0 +1,107 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from pynvml import *
+from multiprocessing.pool import ThreadPool
+import time
+
+from model_analyzer.monitor.record.memory import MemoryRecord
+from model_analyzer.monitor.record.collector import RecordCollector
+from model_analyzer.monitor import Monitor
+
+
+class NVMLMonitor(Monitor):
+    """Use NVML to monitor GPU metrics
+    """
+
+    def __init__(self, frequency):
+        """
+        Parameters
+        ----------
+        frequency : int
+            Sampling frequency for the metric
+        """
+        super().__init__(frequency)
+        nvmlInit()
+
+        # FIXME Handle the case when the number of GPUs is larger than one and
+        # CUDA_VISIBLE_DEVICES is set
+        self._nvml_handles = []
+        for i, gpu in enumerate(self._gpus):
+            self._nvml_handles.append(nvmlDeviceGetHandleByIndex(i))
+
+        # Is the background thread active
+        self._thread_active = False
+
+        # Background thread collecting results
+        self._thread = None
+
+        # Thread pool
+        self._thread_pool = ThreadPool(processes=1)
+
+    def _monitoring_loop(self, tags):
+        self._thread_active = True
+        frequency = self._frequency
+
+        record_collector = RecordCollector()
+        while self._thread_active:
+            begin = time.time()
+            for tag in tags:
+                if tag == 'memory':
+                    for handle in self._nvml_handles:
+                        memory = nvmlDeviceGetMemoryInfo(handle)
+                        memory_record = MemoryRecord(memory.used, memory.total,
+                                                     memory.free)
+                        record_collector.insert(memory_record)
+
+            duration = time.time() - begin
+            if duration < frequency:
+                time.sleep(frequency - duration)
+
+        return record_collector
+
+    def start_recording_metrics(self, tags):
+        """Start recording a list of tags
+
+        Parameters
+        ----------
+        tags : list
+            Name of the metrics to be collected. Options include
+            *memory* and *compute*.
+        """
+        self._thread = self._thread_pool.apply_async(self._monitoring_loop,
+                                                     (tags, ))
+
+    def stop_recording_metrics(self):
+        """Stop recording metrics
+
+        Returns
+        -------
+        RecordCollector
+            A RecordCollector containing all the results
+        """
+        self._thread_active = False
+        record_collector = self._thread.get()
+        self._thread = None
+
+        return record_collector
+
+    def destroy(self):
```

Comment:
Maybe this is a best practice, but is there any other reason why we're letting the caller deallocate the resources?

Reply:
The reason I added this is the warning below from Python documentation:

```
Warning multiprocessing.pool objects have internal resources that need to be properly managed (like any other resource) by using the pool as a context manager or by calling close() and terminate() manually. Failure to do this can lead to the process hanging on finalization.
Note that is not correct to rely on the garbage colletor to destroy the pool as CPython does not assure that the finalizer of the pool will be called (see object.__del__() for more information).
```

We can probably use greenlet threads too, but I used ThreadPools so that monitoring doesn't influence the execution of the main thread.

Reply:
Would it not be enough to override `NVLMonitor.__del__()` and free the resources there?

Reply:
I wanted to do that. But it seems like it is not guaranteed to be called:

```
 CPython does not assure that the finalizer of the pool will be called (see object.__del__() for more information).
```

Reply:
Ah okay this makes sense. The garbage collector isn't guaranteed to run right away.